tests/unit/test_unique_any.cpp

```diff
@@ -0,0 +1,394 @@
+
+    // test that exceptions are thrown for invalid casts
+    EXPECT_EQ(util::any_cast<int>(&as), nullptr);
+    EXPECT_EQ(util::any_cast<std::string>(&ai), nullptr);
+    unique_any empty;
+    EXPECT_EQ(util::any_cast<int>(&empty), nullptr);
+    EXPECT_EQ(util::any_cast<int>((util::unique_any*)nullptr), nullptr);
+
+    // Check that constness of the returned pointer matches that the input.
+    {
+        unique_any a(42);
+        auto p = util::any_cast<int>(&a);
+        static_assert(std::is_same<int*, decltype(p)>::value,
+                "any_cast(any*) should not return const*");
+    }
+    {
+        const unique_any a(42);
+        auto p = util::any_cast<int>(&a);
+        static_assert(std::is_same<const int*, decltype(p)>::value,
+                "any_cast(const any*) should return const*");
+    }
+}
+
+// test any_cast(unique_any&)
+TEST(unique_any, any_cast_ref) {
+    util::unique_any ai(42);
+    auto& i = util::any_cast<int&>(ai);
+
+    EXPECT_EQ(typeid(i), typeid(int));
+    EXPECT_EQ(i, 42);
+
+    // any_cast<T>(unique_any&) returns a 
+    i = 100;
+    EXPECT_EQ(util::any_cast<int>(ai), 100);
+}
+
+// test any_cast(const unique_any&)
+TEST(unique_any, any_cast_const_ref) {
+    const util::unique_any ai(42);
+    auto& i = util::any_cast<const int&>(ai);
+
+    EXPECT_EQ(typeid(i), typeid(int));
+    EXPECT_EQ(i, 42);
+
+    EXPECT_TRUE((std::is_same<const int&, decltype(i)>::value));
+}
+
+// test any_cast(any&&)
+TEST(unique_any, any_cast_rvalue) {
+    auto moved = util::any_cast<moveable>(util::unique_any(moveable()));
+    EXPECT_EQ(moved.moves, 2);
+    EXPECT_EQ(moved.copies, 0);
+}
+
+TEST(unique_any, std_swap) {
+    util::unique_any a1(42);
+    util::unique_any a2(3.14);
+
+    auto pi = util::any_cast<int>(&a1);
+    auto pd = util::any_cast<double>(&a2);
+
+    std::swap(a1, a2);
+
+    // test that values were swapped
+    EXPECT_EQ(util::any_cast<int>(a2), 42);
+    EXPECT_EQ(util::any_cast<double>(a1), 3.14);
+
+    // test that underlying pointers did not change
+    EXPECT_EQ(pi, util::any_cast<int>(&a2));
+    EXPECT_EQ(pd, util::any_cast<double>(&a1));
+}
+
+// test operator=(any&&)
+TEST(unique_any, assignment_from_rvalue) {
+    using util::unique_any;
+    using std::string;
+
+    auto str1 = string("one");
+    auto str2 = string("two");
+    unique_any a(str1);
+
+    unique_any b;
+    b = std::move(a); // move assignment
+
+    EXPECT_EQ(str1, util::any_cast<string>(b));
+
+    EXPECT_EQ(nullptr, util::any_cast<string>(&a));
+}
+
+// test template<typename T> operator=(T&&)
+TEST(unique_any, assignment_from_value) {
+    using util::unique_any;
+    std::vector<int> tmp{1, 2, 3};
+
+    // take a pointer to the orignal data to later verify
+    // that the value was moved, and not copied.
+    auto ptr = tmp.data();
+
+    unique_any a;
+    a = std::move(tmp);
+
+    auto vec = util::any_cast<std::vector<int>>(&a);
+
+    // ensure the value was moved
+    EXPECT_EQ(ptr, vec->data());
```

Comment:
This guarantee is a little hard to validate against the standard: it _is_ guaranteed, but requires knowing that the standard allocator has `propagate_on_container_move_assignment::value` true and that this in turn indicates that pointers to elements are not invalidated on move (clarified in http://www.open-std.org/JTC1/SC22/WG21/docs/lwg-active.html#2321).

I think it would be clearer to use e.g. `testing::nocopy<T>` and check the number of move-assign and move-ctor invokations.